lib/regexpcache/regexpcache.go

```diff
@@ -0,0 +1,201 @@
+package regexpcache
+
+import (
+	"container/heap"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fasttime"
+)
+
+// Cache caches any values with limited number of entries
+// evicted in LRU order.
+// Cache must be created only via NewCache function
+type Cache struct {
+	requests          uint64
+	misses            uint64
+	cleanerMustStopCh chan struct{}
+	cleanerStoppedCh  chan struct{}
+	mu                sync.Mutex
+	m                 map[string]*cacheEntry
+	perKeyMisses      map[string]int
+	getMaxSize        func() int
+	lah               lastAccessHeap
+}
+
+type cacheEntry struct {
+	lastAccessTime uint64
+	heapIdx        int
+
+	key   string
+	value interface{}
+}
+
+// NewCache returns new Cache with limited entries count
+func NewCache(getMaxSize func() int) *Cache {
+	c := &Cache{
+		m:                 make(map[string]*cacheEntry),
+		perKeyMisses:      make(map[string]int),
+		getMaxSize:        getMaxSize,
+		cleanerStoppedCh:  make(chan struct{}),
+		cleanerMustStopCh: make(chan struct{}),
+	}
+	go c.cleaner()
+	return c
+}
+
+// MustStop frees up resources occupied by c.
+func (c *Cache) MustStop() {
+	close(c.cleanerMustStopCh)
+	<-c.cleanerStoppedCh
+}
+
+func (c *Cache) cleaner() {
+	ticker := time.NewTicker(55 * time.Second)
+	defer ticker.Stop()
+	perKeyMissesTicker := time.NewTicker(4 * time.Minute)
+	defer perKeyMissesTicker.Stop()
+	for {
+		select {
+		case <-c.cleanerMustStopCh:
+			close(c.cleanerStoppedCh)
+			return
+		case <-ticker.C:
+			c.cleanByTimeout()
+		case <-perKeyMissesTicker.C:
+			c.cleanPerKeyMisses()
```

Comment:
I don't think perKeyMisses optimization is needed here contrary to `lib/blockcache`, since it is unlikely to have many queries with one-off regexps.

Reply:
Removed